python/ray/dag/dag_node_operation.py

```diff
@@ -597,9 +607,13 @@ def _generate_overlapped_execution_schedule(
         # Swap each NCCL read with the previous compute to overlap the NCCL read
         # with computation.
         for i in range(1, len(overlapped_schedule)):
-            if (
+            if not overlapped_schedule[i - 1].requires_nccl_op and (
                 overlapped_schedule[i].requires_nccl_read
-                and not overlapped_schedule[i - 1].requires_nccl_op
+                or (
+                    overlapped_schedule[i].requires_nccl_collective
```

Comment:
For DDP, the schedule is `backward(i) -> collective(i) -> update(i)` and `backward(i) -> backward(i-1)`. Does this check overlap what we want, namely `backward(i-1)` and `collective(i)` should be parallel.

Reply:
Overlapping collective operations and communications is in alpha stage. They are overlapped by prioritizing NCCL operations over non-NCCL operations in the `__lt__` function of the `_DAGOperationGraphNode`.